<center><h1><b>E-Commerce Analysis</b></h1></center>

### Overview

The e-commerce industry has revolutionized modern commerce, transforming how customers shop and businesses operate. What once seemed impossible, such as ordering products and receiving them at home within minutes, has become reality through sophisticated e-commerce platforms like Flipkart, Amazon, and others. This transformation is made possible through the strategic application of **data science and analytics**, which powers everything from recommendation systems to supply chain optimization and customer behavior analysis.

E-commerce companies generate massive amounts of data daily from millions of customers shopping across various platforms. Both **online and offline retail operations** contribute to this data ecosystem, from digital platforms to physical stores of major retail chains like Big Bazaar, Croma, Lifestyle, and D-Mart. The ability to properly analyze and utilize this data is crucial for companies to make informed business decisions, improve strategies, and drive revenue growth.

**Sales analysis** is particularly critical across all industries, whether in healthcare, finance, AI, e-commerce, or retail. Every company, regardless of sector, relies on sales performance to measure success and plan for growth. Understanding sales patterns, profit margins, customer segments, and product performance enables businesses to optimize their operations and maximize profitability.

In this project, I focused on analyzing e-commerce sales data from both online and offline channels, providing insights that would be valuable for my role as a data analyst, business analyst, or data scientist. The dataset contains comprehensive information including order details, shipping modes, customer demographics, product categories, quantities, sales figures, discounts, and profit margins, representing the type of real-world data that drives business intelligence in modern retail operations.

As of 2024, the e-commerce sector continues to expand globally, with companies investing heavily in **data science, machine learning, and analytics** to build robust recommendation engines, optimize inventory management, and enhance customer experiences. This project addresses the practical challenges and opportunities that arise when analyzing such comprehensive sales data.

---

### Objectives

1. Calculate the monthly sales of the store and identify which month had the highest sales and which month had the lowest sales.
2. Analyze sales based on product categories and determine which category has the lowest sales and which category has the highest sales.
3. Conduct sales analysis based on sub-categories to understand performance at a granular level.
4. Analyze the monthly profit from sales and determine which month had the highest profit.
5. Examine profit by category and sub-category to identify the most and least profitable segments.
6. Analyze sales and profit by customer segment to understand different customer demographics and their purchasing patterns.
7. Calculate and analyze the sales-to-profit ratio to evaluate the efficiency and profitability of sales operations.

Through this comprehensive analysis, I aim to provide actionable insights that can help e-commerce businesses optimize their operations, improve profitability, and make data-driven strategic decisions.

### Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
import plotly.colors as colors
import warnings
warnings.filterwarnings("ignore")
%matplotlib inline

Setting the default template as white

In [2]:
pio.templates.default = 'plotly_white'

Reading the csv file into a pandas dataframe

In [ ]:
df = pd.read_csv('Superstore.csv') 

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 2944: invalid start byte

Since the data contains slash and dash, I got an error while reading the csv file. So I used `encoding='latin1'` to read the csv file

In [4]:
df = pd.read_csv('Superstore.csv', encoding='latin1')
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


Checking the data types of the columns

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

Checking for null values

In [9]:
df.isnull().sum().sum().item(0)

0

There is no null values in the dataset

Checking for any duplicate values

In [11]:
df.duplicated().sum().item(0)

0

There are no duplicate values in the dataset

Checking the number of rows and columns in the dataset

In [12]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 9994
Number of columns: 21


The e-commerce dataset contains almost 10K rows for each entry and 21 features for each entry.

Getting a description of the data using the describe() function on the numeric columns

In [5]:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


I can get a lot of information from just this descriptive statistics table. The average sales was $230, with a standard deviation of $623. This means that there is a lot of variation in the sales. The maximum sales was $22638. Similarly, the average quantity was almost 3.8, average discount was 15.6%, and the average profit was $28.66. The maximum profit was $8400, whereas the minimum profit was -$6600. The maximum quantity was 14 and the maximum discount was 80%.

Changing the data type of the "Order Date" and "Ship Date" columns from object to datetime.

In [14]:
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

Since I want to analyze monthly and yearly trends, I need to extract the year and month from the "Order Date" column.

In [15]:
df["Order_Year"] = df["Order Date"].dt.year
df["Order_Month"] = df["Order Date"].dt.month
df["Order_Day"] = df["Order Date"].dt.dayofweek
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Order_Year,Order_Month,Order_Day
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,2016,11,1
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,2016,11,1
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,2016,6,6
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,2015,10,6
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,2015,10,6


Dropping the "Row ID" column as it is not needed for analysis.

In [16]:
df.drop("Row ID", axis=1, inplace=True)

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Order ID       9994 non-null   object        
 1   Order Date     9994 non-null   datetime64[ns]
 2   Ship Date      9994 non-null   datetime64[ns]
 3   Ship Mode      9994 non-null   object        
 4   Customer ID    9994 non-null   object        
 5   Customer Name  9994 non-null   object        
 6   Segment        9994 non-null   object        
 7   Country        9994 non-null   object        
 8   City           9994 non-null   object        
 9   State          9994 non-null   object        
 10  Postal Code    9994 non-null   int64         
 11  Region         9994 non-null   object        
 12  Product ID     9994 non-null   object        
 13  Category       9994 non-null   object        
 14  Sub-Category   9994 non-null   object        
 15  Product Name   9994 n

---
### Monthly Sales Analysis

For monthly sales analysis, I grouped the data by the "Order_Month" column and calculated the sum of the "Sales" column for each month. Then, I sorted the results in descending order and reset the index.

In [22]:
sales_by_month = df.groupby("Order_Month")["Sales"].sum().reset_index()
sales_by_month.sort_values(by="Sales", ascending=False, inplace=True)
sales_by_month.reset_index(drop=True, inplace=True)
sales_by_month

,Order_Month,Sales
0,11,352461.0710
1,12,325293.5035
2,9,307649.9457
3,3,205005.4888
4,10,200322.9847
5,8,159044.0630
6,5,155028.8117
7,6,152718.6793
8,7,147238.0970
9,4,137762.1286


November is the month with the highest sales of 352461 units, followed by December and September. However, February and January have the lowest sales.

Creating a line chart to visualize the sales trend over time.

In [25]:
fig = px.line(sales_by_month.sort_values('Order_Month'), x='Order_Month', y='Sales', title='Sales by Month')
fig.show()

Line chart makes it easy to see the trend of sales over time. As we can see that sales increase from February (59.7k) to March (205k) and then drops a bit in April (138k). The sales remain stable from May to August at 150k. Then, sales largley increases in September (307k) and then drops in October (200k). However, it reaches the highest sales in November (352k) and remains high in December.

Hence, the sale is highest in November and lowest in February.

---

### Analyzing Sales based on product categories

In [29]:
sales_by_category = df.groupby('Category')['Sales'].sum().reset_index()
sales_by_category.sort_values(by='Sales', ascending=False, inplace=True)
sales_by_category.reset_index(drop=True, inplace=True)
sales_by_category

,Category,Sales
0,Technology,836154.0330
1,Furniture,741999.7953
2,Office Supplies,719047.0320


There are only 3 categories in the dataset. `Technology` category has the highest sales of 836,154, followed by `Furniture` and `Office Supplied` categories.

Creating a pie chart to show the sales of each category

In [31]:
fig = px.pie(sales_by_category, values='Sales', names='Category', title='Sales by Category')
fig.show()

As we can see from the pie chart, the `Technology` category consists of 36.4%, followed by `Furniture` and `Office Supplied` categories with sales of 32.3% and 31.3% respectively.

Creating a same pie chart but in a fancy way with a hole in the middle and changing the color palette.

In [38]:
fig = px.pie(sales_by_category, 
values='Sales', 
names='Category', 
title='Sales by Category',
color_discrete_sequence=px.colors.qualitative.Pastel)

fig.update_traces(hole=.4, 
textinfo='label+percent', 
textposition='inside',
marker=dict(colors=px.colors.qualitative.Pastel))

fig.update_layout(title_font_size=20)

fig.show()

---

### Analyzing Sales by Sub Category

In [41]:
sales_by_subcategory = df.groupby('Sub-Category')['Sales'].sum().reset_index()
sales_by_subcategory.sort_values(by='Sales', ascending=False, inplace=True)
sales_by_subcategory.reset_index(drop=True, inplace=True)
sales_by_subcategory

,Sub-Category,Sales
0,Phones,330007.0540
1,Chairs,328449.1030
2,Storage,223843.6080
3,Tables,206965.5320
4,Binders,203412.7330
5,Machines,189238.6310
6,Accessories,167380.3180
7,Copiers,149528.0300
8,Bookcases,114879.9963
9,Appliances,107532.1610


There are 16 sub-categories in the dataset. `Phones` sub-category has the highest sales of 330,007, followed by `Chairs` and `Storage`. `Fasteners` sub-category has the lowest sales of 3024 only.

Creating a bar chart to show the sales of each sub-category

In [43]:
fig = px.bar(sales_by_subcategory, 
x='Sub-Category', 
y='Sales', 
title='Sales by Sub-Category')
fig.show()

In [83]:
print("Phones % of sales: ", float(sales_by_subcategory[sales_by_subcategory["Sub-Category"] == "Phones"]["Sales"]/sales_by_subcategory["Sales"].sum() * 100))
print("Fasteners % of sales: ", float(sales_by_subcategory[sales_by_subcategory["Sub-Category"] == "Fasteners"]["Sales"]/sales_by_subcategory["Sales"].sum() * 100))

Phones % of sales:  14.365615985225741
Fasteners % of sales:  0.13165065590324776


In [50]:
print("There are 16 sub-categories in the dataset.")
print("Phones sub-category has the highest sales which comes under", df[df["Sub-Category"] == "Phones"]["Category"].values[0], "category", "with sales of", sales_by_subcategory[sales_by_subcategory["Sub-Category"] == "Phones"]["Sales"].values[0])
print("Fasteners sub-category has the lowest sales which comes under", df[df["Sub-Category"] == "Fasteners"]["Category"].values[0], "category", "with sales of", sales_by_subcategory[sales_by_subcategory["Sub-Category"] == "Fasteners"]["Sales"].values[0])

There are 16 sub-categories in the dataset.
Phones sub-category has the highest sales which comes under Technology category with sales of 330007.054
Fasteners sub-category has the lowest sales which comes under Office Supplies category with sales of 3024.28


---

### Analyzing the Monthly profit by category

In [58]:
profit_by_month = df.groupby("Order_Month")["Profit"].sum().reset_index()
profit_by_month.sort_values(by="Profit", ascending=False, inplace=True)
profit_by_month.reset_index(drop=True, inplace=True)
profit_by_month

,Order_Month,Profit
0,12,43369.1919
1,9,36857.4753
2,11,35468.4265
3,10,31784.0413
4,3,28594.6872
5,5,22411.3078
6,8,21776.9384
7,6,21285.7954
8,7,13832.6648
9,4,11587.4363


As we can see, even though `November` has the highest sales, `December` has the highest profit, followed by `September` and then `November`. However, `January` and `February` have the lowest profit because they have the lowest sales.

Creating a bar chart and line chart to show the profit by month

In [59]:
fig = px.bar(profit_by_month, x="Order_Month", y="Profit", title="Profit by Month")
fig.show()

In [61]:
fig = px.line(profit_by_month.sort_values(by="Order_Month", ascending=True), x="Order_Month", y="Profit", title="Profit by Month")
fig.show()

We can clearly see that `December` has the highest profit and `January` has the lowest profit. The profit increased in `March` but decreased rapidly in `April`. The profit then increased agin and decreased. It started to steadily increase from `August` to `December`.

Calculating the average profit per month

In [55]:
profit_by_month_avg = df.groupby("Order_Month")["Profit"].mean().reset_index()
profit_by_month_avg.sort_values(by="Profit", ascending=False, inplace=True)
profit_by_month_avg.reset_index(drop=True, inplace=True)
profit_by_month_avg

,Order_Month,Profit
0,3,41.084321
1,10,38.808353
2,2,34.315369
3,8,30.845522
4,12,30.801983
5,5,30.491575
6,6,29.687302
7,9,26.650380
8,11,24.111779
9,1,23.974924


---

### Analyzing Profit by Category and Sub-Category

In [62]:
profit_by_category = df.groupby('Category')['Profit'].sum().reset_index()
profit_by_category.sort_values(by='Profit', ascending=False, inplace=True)
profit_by_category.reset_index(drop=True, inplace=True)
profit_by_category


,Category,Profit
0,Technology,145454.9481
1,Office Supplies,122490.8008
2,Furniture,18451.2728


As we can see, `Technology` has the highest profit and the highest sales. However, even though `Furniture` has little higher sales than `Office Supplies`, `Office Supplies` had much larger profit margin than `Furniture`. Hence, company should focus on `Office Supplies` to increase profit margin even though it has lower sales than `Furniture`.

Creating a pie chart to show the profit margin of each category

In [63]:
fig = px.pie(profit_by_category, 
values='Profit', 
names='Category', 
title='Profit by Category',
color_discrete_sequence=px.colors.qualitative.Pastel)

fig.update_traces(hole=.4, 
textinfo='label+percent', 
textposition='inside',
marker=dict(colors=px.colors.qualitative.Pastel))

fig.update_layout(title_font_size=20)

fig.show()

As we can see, `Technology` contributes 50.8% of the total profit with a sales of 36.4%, while `Furniture` contributes 6.44% of the total profit with a sales of 32.3%, `Office Supplies` contributes 42.8% of the total profit with a sales of 31.3%.

Now, calculating the profit by sub category

In [64]:
profit_by_subcategory = df.groupby('Sub-Category')['Profit'].sum().reset_index()
profit_by_subcategory = profit_by_subcategory.sort_values(by='Profit', ascending=False)
profit_by_subcategory.reset_index(drop=True, inplace=True)
profit_by_subcategory

,Sub-Category,Profit
0,Copiers,55617.8249
1,Phones,44515.7306
2,Accessories,41936.6357
3,Paper,34053.5693
4,Binders,30221.7633
5,Chairs,26590.1663
6,Storage,21278.8264
7,Appliances,18138.0054
8,Furnishings,13059.1436
9,Envelopes,6964.1767


`Copiers` sub-category has the highest profit of 55,618, followed by `Phones` which has the highest sales and `Accessories`. However, `Tables` sub-category has the lowest profit of -17,725, followed by `Bookcases` and `Supplies`. This could be because this subcategory are furtinures which require a high cost of materials and labor to make them which make them lose money. However, `Fasteners` which had the lowest sales has a positive profit of 950 which is better than others where the sub category is lossing money.

In [86]:
print("Copiers with highest profit come under", df[df["Sub-Category"] == "Copiers"]["Category"].values[0])
print("Tables with lowest profit come under", df[df["Sub-Category"] == "Tables"]["Category"].values[0])

Copiers with highest profit come under Technology
Tables with lowest profit come under Furniture


Creating a bar chart to show the profit of each sub-category

In [65]:
fig = px.bar(profit_by_subcategory, x='Sub-Category', y='Profit', title='Profit by Sub-Category')
fig.show()

---
### Analyzing sales and profit by customer segment

In [66]:
customer_segment = df.groupby('Segment').agg({'Sales': 'sum', 'Profit': 'sum'}).reset_index()
customer_segment.sort_values(by='Profit', ascending=False, inplace=True)
customer_segment.reset_index(drop=True, inplace=True)
customer_segment

,Segment,Sales,Profit
0,Consumer,1.161401e+06,134119.2092
1,Corporate,7.061464e+05,91979.1340
2,Home Office,4.296531e+05,60298.6785


There are 3 customer segments. `Consumer` segment has the highest profit and sales of 134,119 and 1,161,401, followed by `Corporate` and `Home Office`.


Creating a double bar chart to compare sales and profit of each customer segment.

In [71]:
fig = go.Figure()

fig.add_trace(go.Bar(x=customer_segment['Segment'], 
y=customer_segment['Sales'], 
name='Sales', 
marker_color = colors.qualitative.Pastel[0]))

fig.add_trace(go.Bar(x=customer_segment['Segment'], 
y=customer_segment['Profit'], 
name='Profit', 
marker_color= colors.qualitative.Pastel[1]))

fig.update_layout(
    title='Sales and Profit by Customer Segment',
    xaxis_title='Customer Segment',
    yaxis_title='Amount',
    barmode='group'
)

fig.show()


As we can clearly see from the bar chart, Consumer has the highest sales and profit, followed by Corporate and Home Office.

---
### Analyzing sales to profit ratio

In [74]:
customer_segment['Sales to Profit Ratio'] = customer_segment['Sales'] / customer_segment['Profit']
customer_segment[["Segment", "Sales to Profit Ratio"]].sort_values(by='Sales to Profit Ratio', ascending=False)

,Segment,Sales to Profit Ratio
0,Consumer,8.659471
1,Corporate,7.677245
2,Home Office,7.125416


`Consumer` segment has the highest sales to profit ratio of 8.66, followed by `Corporate` and `Home Office` segments. This indicates that the `Consumer` segment generates the most profit relative to its sales.

Visualize the sales to profit ratio for each segment using a bar chart.

In [77]:
fig = px.bar(customer_segment, x='Segment', y='Sales to Profit Ratio', title='Sales to Profit Ratio by Segment')
fig.show()

## Conclusion:

1. `November` had the highest sales and `February` had the lowest sales.
2. `Technology` category had the highest sales of 36.4% and `Office Supplies` had the lowest sales of 31.3%.
3. `Phones` sub-category which comes under `Technology` category had the highest sales of 14.3% and `Fasteners` which comes under `Office Supplies` had the lowest sales of 0.13%.
4. `December` had the highest profit and `January` had the lowest profit.
5. `Technology` category had the highest profit and `Furniture` had the lowest profit. Similarly, `Copiers` sub-category which comes under `Technology` category had the highest profit and `Tables` which comes under `Furniture` had the lowest profit.
6. `Consumer` segment has the highest profit and sales, followed by `Corporate` and `Home Office` segments.
7. `Consumer` segment has the highest sales to profit ratio, followed by `Corporate` and `Home Office` segments.

---

## Author

- [Parshv Patel](https://www.linkedin.com/in/parshv-patel-65a90326b/)